ffmpeg

eruffaldi opened the pull request:
resolves #10583 

### This pullrequest changes
Module: videoio
File: cap_ffmpeg_impl.hpp
Changes: CvVideoWriter_FFMPEG::open
Location: switch of codecs

eruffaldi commented:
# Doc Build
trailing error (FIXED)

# Linux Build error
undefined AV_PIX_FMT_BGR0 (depends on ffmpeg version) (FIXED)

# Mac Build error
test_videoio for AVI and fourcc IYUV. This fourcc is considered as rawvideo by ffmpeg (riff.c file { AV_CODEC_ID_RAWVIDEO,     MKTAG('I', 'Y', 'U', 'V') }). But IYUV == I420 and in any case the image will be altered using rawvideo

Hypothesis: different ffmpeg version than Windows build

alalek commented:
1.

According to this:
https://github.com/opencv/opencv/blame/3.4.0/modules/videoio/src/cap_ffmpeg_impl.hpp#L1923-L1929

there are only two possible values for "input_pix_fmt". Why we need such large switches?

2.

Check in ffmpeg itself (ffplay or extract frames into png) that result file doesn't swap colors (OpenCV's color format is BGR: `red -> Scalar(0, 0, 255)`)

3.

Squash all fixup commits into one. No need to keep them in the history.

eruffaldi commented:
1. (and 2.) input_pix_fmt only AV_PIX_FMT_BGR24 and AV_PIX_FMT_GRAY8
I acted in this way looking at the original code where the switch was already addressing more than these input formats (e.g. for CODEC_ID_RAWVIDEO):

https://github.com/opencv/opencv/blame/3.4.0/modules/videoio/src/cap_ffmpeg_impl.hpp#L2010-L2013

So I have entered all the input formats that are supported by the specific codec (via ffprobe).

input_pix_fmt should have more values in the case (1) 4 channels  input image, (2) 16-bit single channel input image. This would allow to save alpha channel images and high-depth grayscale images as supported by ffmpeg codecs.

This means two additional changes
1) CvVideoWriter_FFMPEG::writeFrame (easy based on input_pix_fmt)
2) CvVideoWriter_FFMPEG::open (affecting the caller interface)

eruffaldi commented:
Changed:
- added handling of FOURCC(255,255,255,255) that prints the list of FOURCC for the given output format 
- RAWVIDEO codec given FOURCC=RGBA forces pix_fmt = RGBA
- otherwise RAWVIDEO acts as original code

eruffaldi commented:
Fixed White space

addisonElliott commented:
What is stopping this from being merged?

I have an application where I am wanting lossless video I/O. Currently, I am using MJLS codec to have this occur but if I attempt to use rawvideo, it converts to YUV420 and there is some loss in that conversion.

I think the switches containing multiple values are good to have, because it would be nice to expand support for multiple formats in the future.

alalek commented on modules/videoio/src/cap_ffmpeg_impl.hpp:
```diff
@@ -1928,6 +1946,13 @@ bool CvVideoWriter_FFMPEG::open( const char * filename, int fourcc,
 
+    if(fourcc == -1)
+    {
+        fprintf(stderr,"OpenCV: FFMPEG: format %s / %s, codec id %04X\n", fmt->name, fmt->long_name, codec_id);
```
codec_id is always NONE here. Removed
